File: src/MartianRobots.Core/Simulator.cs
namespace MartianRobots.Core;

public class Simulator
{
    private readonly CommandFactory _commandFactory = new();

    public IReadOnlyList<Robot> Run(ProblemInput input)
    {
        var world = new World(input.MaxX, input.MaxY);
        var robots = new List<Robot>();

        foreach (var robotInput in input.Robots)
        {
            var robot = new Robot(
                robotInput.X,
                robotInput.Y,
                robotInput.Orientation);

            foreach (var instruction in robotInput.Instructions)
            {
                var command = _commandFactory.Create(instruction);
                command.Execute(robot, world);

                if (robot.Lost)
                    break;
            }

            robots.Add(robot);
        }

        return robots;
    }
}

File: go-implementation/simulator.go
package main

type RobotInput struct {
	X            int
	Y            int
	Orientation  Orientation
	Instructions string
}

type ProblemInput struct {
	MaxX   int
	MaxY   int
	Robots []RobotInput
}

func Simulate(input ProblemInput) []Robot {
	world := NewWorld(input.MaxX, input.MaxY)
	results := make([]Robot, 0, len(input.Robots))

	for _, inputRobot := range input.Robots {
		robot := Robot{
			X:           inputRobot.X,
			Y:           inputRobot.Y,
			Orientation: inputRobot.Orientation,
		}

		for _, instruction := range inputRobot.Instructions {
			switch instruction {
			case 'L':
				robot.TurnLeft()
			case 'R':
				robot.TurnRight()
			case 'F':
				robot.MoveForward(world)
			}

			if robot.Lost {
				break
			}
		}

		results = append(results, robot)
	}

	return results
}
